src/displaycontroller.h

```diff
@@ -2313,6 +2343,112 @@ class GenericBitmappedDisplayController : public BitmappedDisplayController {
+                                     TRawGetRow rawGetRow, /* TRawGetPixelInRow rawGetPixelInRow, */ TRawSetPixelInRow rawSetPixelInRow)
+  {
+    // transformed bitmap plot works as follows:
+    // 1. for each pixel in the destination rectangle, calculate the corresponding pixel in the source bitmap
+    // 2. if the source pixel is within source, and is not transparent, plot it to the destination
+    //
+    // drawingRect should be all on-screen, pre-clipped, but moved to -originX, -originY
+    float pos[3] = {0.0f, 0.0f, 1.0f};
+    float srcPos[3] = {0.0f, 0.0f, 1.0f};
+    float maxX = drawingRect.X2;
+    float maxY = drawingRect.Y2;
+    auto data = bitmap->data;
+    const int rowlen = (bitmap->width + 7) / 8;
+ 
+    for (float y = drawingRect.Y1; y <= maxY; y++) {
+      for (float x = drawingRect.X1; x <= maxX; x++) {
+        // calculate the source pixel
+        pos[0] = x;
+        pos[1] = y;
+        dspm_mult_f32(invMatrix, pos, srcPos, 3, 3, 1);
+        
+        int srcXint = (int) (srcPos[0] - 0.5f);
+        int srcYint = (int) (srcPos[1] - 0.5f);
+        if (srcXint >= 0 && srcXint < bitmap->width && srcYint >= 0 && srcYint < bitmap->height) {
+
+          auto srcRow = data + srcYint * rowlen;
+          if ((srcRow[srcXint >> 3] << (srcXint & 7)) & 0x80)
+            rawSetPixelInRow(rawGetRow((int)y + originY), (int)x + originX);
+        }
+      }
+    }
+  }
+
+
+  template <typename TRawGetRow, /*typename TRawGetPixelInRow,*/ typename TRawSetPixelInRow /*, typename TBackground */>
+  void genericRawDrawTransformedBitmap_RGBA2222(int originX, int originY, Rect drawingRect, Bitmap const * bitmap, const float * invMatrix,
+                                     TRawGetRow rawGetRow, /* TRawGetPixelInRow rawGetPixelInRow, */ TRawSetPixelInRow rawSetPixelInRow)
+  {
+    // transformed bitmap plot works as follows:
+    // 1. for each pixel in the destination rectangle, calculate the corresponding pixel in the source bitmap
+    // 2. if the source pixel is within source, and is not transparent, plot it to the destination
+    //
+    // drawingRect should be all on-screen, pre-clipped, but moved to -originX, -originY
+    float pos[3] = {0.0f, 0.0f, 1.0f};
+    float srcPos[3] = {0.0f, 0.0f, 1.0f};
+    float maxX = drawingRect.X2;
+    float maxY = drawingRect.Y2;
+    auto data = bitmap->data;
+    const int width = bitmap->width;
+ 
+    for (float y = drawingRect.Y1; y <= maxY; y++) {
+      for (float x = drawingRect.X1; x <= maxX; x++) {
+        // calculate the source pixel
+        pos[0] = x;
+        pos[1] = y;
+        dspm_mult_f32(invMatrix, pos, srcPos, 3, 3, 1);
+        
+        int srcXint = (int) (srcPos[0] - 0.5f);
+        int srcYint = (int) (srcPos[1] - 0.5f);
+        if (srcXint >= 0 && srcXint < bitmap->width && srcYint >= 0 && srcYint < bitmap->height) {
```

Comment:
these routines, and others, do some tweaking of position values for integer rounding - I'm not 100% certain this is all correct/sensible

there may also be some scope for some optimisation here to avoid converting the calculated integer source pixel positions within the bitmap, and instead have the `if` statement use floating point values for comparison